e2e/_suites/ingest-manager/features/agent_endpoint_integration.feature

```diff
@@ -0,0 +1,38 @@
+@agent_endpoint_integration
+Feature: Agent Endpoint Integration
+  Scenarios for Agent to deploy Endpoint and sending data to Ingest Manager and Elasticsearch.
+
+@deploy-endpoint-with-agent
+Scenario: Adding the Endpoint Integration to an Agent makes the host to show in Security App
+  Given a "centos" agent is deployed to Fleet
+  When the "latest" version of the "Endpoint" package is installed
+    And the "Endpoint" integration is "added" in the "default" configuration
+  Then the "Endpoint" datasource is shown in the "default" configuration as added
+#    And the host name is shown in the Security App
```

Comment:
I have the info we need, and suggestions I'd recommend for us to finish this off.

item 1) We really should parameterize the 2 calls that relate to Agent or Endpoint being online or offline.  So, I recommend we change
  Then the agent is not listed as online in Fleet
to
  Then the agent is listed as "offline" in Fleet 
with support also for the opposite:
  Then the agent is listed as "online" in Fleet 

And we'd change:
    And the endpoint is not listed in Security App as online
to
    And the endpoint is listed as "offline" in Security App
along with the opposite:
    And the endpoint is listed as "online" in Security App

Sound good?

For the step above about 'host name shown in security app' I think it makes sense to come after a call to ensure the Endpoint is listed as online.

For the API calls that support those queries and assertions, below is how we get the response:

POST to api/endpoint/metadata
(with empty body is fine)
 Then we have to walk the list of responses to find the element with the current Agent ID, which we 
 should know (or can get, right?).  That Agent id is located in the response at, for example:
{
    "hosts": [
        {
            "metadata": {
                "elastic": {
                    "agent": {
                        "id": "a6e59bc1-6252-4a92-95dd-f49787a9c07b"

Note, unless we implement a force-unenroll call for each found Agent before the test, we can't assume it will be the first item in the list above ^

-- then after we find the Agent id, we can query the 'host_status' field at the same level as the metadata field in that list element.  It should be 'online'!

            "host_status": "online"

The host name is under metadata: 
host:hostname:value

I'd like to query a policy status, also under metadata as:
metadata:Endpoint:policy:applied:status: value.  and the value should be 'success'
Should / can we add an explicit feature step here as:
And the endpoint policy status is "success"

This is a great first set of tests for Endpoint, I'm so excited.

screenshot:
<img width="682" alt="endpoint-metadata-api-call" src="https://user-images.githubusercontent.com/12970373/91353128-d4e50900-e7b8-11ea-9726-59e928452eb2.png">


Note: There is a GET api call we can make that is slightly more direct, but we need to know the Endpoint ID first, which I don't believe we have at this point.  Regardless, if helpful later, that call is:
GET to api/endpoint/metadata/[endpoint id]
and it returns the same data as for its list element in the prior list call.

Let me know any thoughts

Reply:
@mdelapenya item # 1) for step:
  Then the policy response will be shown in the Security App

We could re-use the endpoint/metadata call POST, but I would like to exercise another API endpoint that is more direct:
GET to api/endpoint/metadata/[endpoint id]

The endpoint id can be captured from the api/endpoint/metadata call we had just made (or we can make it again)

POST to api/endpoint/metadata
Then we have to walk the list of responses to find the list element with the current Agent ID.  Then after we find the list element for our Agent id, we can look at the rest of that node of the response.  Policy status is under metadata in that node, as:

metadata:Endpoint:policy:applied:status:"success" 
 ... see the screenshot above for more detailed pic

I'd like to add an explicit feature file step here as:
And the endpoint policy status is "success"

Reply:
--------------------------------
@mdelapenya for the other items lets discuss.  1: 
When the policy is updated to have malware in detect mode         

I posted all of the api for this in the other issue:
https://github.com/elastic/e2e-testing/pull/195#discussion_r473282601

I do however have a 'dynamic' complication to ask us to accommodate, so the test steps can be re-used.  can we make 2 inputs in the step variables, as:
When policy is updated with "[original text]" to "[new text]"
 - if we allow the explicit citation of the json we can avoid having to hardcode what to do with those inputs in the go file... if we do this it will be just a find and replace, right?   but it will be easier to implement other policy tests, as we'll undoubtedly want.  I'm invisioning it now as:
When policy is updated with "{"malware":{"mode":"prevent"}" to "{"malware":{"mode":"detect"}"

I hope the quoting is easy to manage in Go.  Single quotes ftw?

In the implementation, we'll need to capture the current policy (config), and do the find & replace, and then push it all back in with the prior noted PUT call.  We can cite to test designers that we intend to change all instances found of the original text, just to be easier on us.  It should be fine

Note that Malware mode is set to 'prevent' by default so it will work the first time we run the test.  But, if we are at all retaining state in the Kibana container, then we should put in a check first, to see if it is already set.  We could check for the 'new text' string and if it is found in the resp, then there is nothing more to do.  

for step 2:
 Then the policy will reflect the change in the Security App

I think we'll need to update it to be dynamic as well, to allow for the assertion...
 Then the integration policy reflects the change "[new text]"

for this case it would then be:
 Then the integration policy reflects the change "{"malware":{"mode":"detect"}"

We can call the noted GET api after we PUT the update and do a match to find the same "new text" string in the resp.   This is a pretty weak assertion tho, as the Security App doesn't currently have as much real time tracking of policy application state as we want, this is more of a CRUD validation and isn't inclusive of any state from the Endpoint app on the host.  The policy application on the host is eventually returned in the Ingest Manager calls for the given Agent in the Activity list (in 10 to 20 or more seconds).  I'd like to do this deeper validation now, if we can.  Here is what API I'd call:

/api/ingest_manager/fleet/agents/4dc9de9d-a349-4579-8142-b84ff882c931/events?page=1&perPage=5
 - the uuid is the Agent ID above, which we should have.  

This will return up to 5 rows of data about the goings on of the Agent.
<img width="1231" alt="Screen Shot 2020-08-31 at 2 46 42 PM" src="https://user-images.githubusercontent.com/12970373/91758443-089ba680-eb9e-11ea-90b7-11675076cf6b.png">

The API response itself, and the row we're looking for is this:
<img width="1297" alt="Screen Shot 2020-08-31 at 3 22 33 PM" src="https://user-images.githubusercontent.com/12970373/91758418-fde11180-eb9d-11ea-832e-2ddce3e22dfc.png">

... So, we can get the events response, and query it for an entry where the message field contains a few keywords: 'endpoint-security' AND 'RUNNING' AND 'policy' as shown in this example:
message: "Application: endpoint-security--7.9.0[4dc9de9d-a349-4579-8142-b84ff882c931]: State changed to RUNNING: Protecting with policy {be6ea1e0-ebb9-11ea-896c-2d695a78748d}"
 -AND we can confirm the timestamp is AFTER the time that the policy change was sent we can accept it.  If we don't mind the timestamp, we might be asserting against the original message sent when Endpoint is started up.  The wording is the same regardless of what is changed in the protection policy, fyi.

I think this is a solid completion to this section of test addition.

Reply:
>I do however have a 'dynamic' complication to ask us to accommodate, so the test steps can be re-used. can we make 2 inputs in the step variables, as:
When policy is updated with "[original text]" to "[new text]"

>if we allow the explicit citation of the json we can avoid having to hardcode what to do with those inputs in the go file... if we do this it will be just a find and replace, right? but it will be easier to implement other policy tests, as we'll undoubtedly want. I'm invisioning it now as:
When policy is updated with "{"malware":{"mode":"prevent"}" to "{"malware":{"mode":"detect"}"

Because we want to use a business language, I'd like to use something like this, hiding any reference to JSON notation in the specs (unless it's a strong  requirements in the specification):

```gherkin
When policy is updated from "malware" in "prevent" mode to "detect" mode
```

My questions for that JSON representation (`{"malware":{"mode":"prevent"}`) are:
1. malware: I guess there are tens of this type
2. mode: are there other keywords here?
3. prevent: the mode type, being possible to have `prevent, detect`, possibly more.

For 2) if we only have modes, I'd keep the step as the above, otherwise I'd use, which captures the dynamic behaviour of the three components:
```gherkin
When policy is updated from "malware" in "prevent" "mode" to "detect"
```
Again for 2) Can we update a policy from `{"malware":{"mode":"prevent"}` to something else `{"malware":{"FOO":"BAR"}` ? Said differently: if we update the mode, do we update the mode and nothing else?

If the second, I'd use this, which expresses this nature of changing the mode to something else (potentially to another mode):
```gherkin
When policy is updated from "malware" in "prevent" "mode" to "detect" "mode"
```

Reply:
Related to the policy and the different modes, I'm hitting this URL: `http://localhost:5601/api/ingest_manager/package_policies/014932d0-eb8a-11ea-b35e-992973a54144`, which returns this payload:

```json
{
    "item": {
        "id": "014932d0-eb8a-11ea-b35e-992973a54144",
        "version": "WzM5MSwxXQ==",
        "name": "Elastic Endpoint",
        "description": "",
        "namespace": "default",
        "policy_id": "f0d1dcf0-eb83-11ea-b35e-992973a54144",
        "enabled": true,
        "output_id": "",
        "inputs": [
            {
                "type": "endpoint",
                "enabled": true,
                "streams": [],
                "config": {
                    "artifact_manifest": {
                        "value": {
                            "manifest_version": "1.0.0",
                            "schema_version": "v1",
                            "artifacts": {
                                "endpoint-exceptionlist-macos-v1": {
                                    "encryption_algorithm": "none",
                                    "decoded_sha256": "d801aa1fb7ddcc330a5e3173372ea6af4a3d08ec58074478e85aa5603e926658",
                                    "decoded_size": 14,
                                    "encoded_sha256": "f8e6afa1d5662f5b37f83337af774b5785b5b7f1daee08b7b00c2d6813874cda",
                                    "encoded_size": 22,
                                    "relative_url": "/api/endpoint/artifacts/download/endpoint-exceptionlist-macos-v1/d801aa1fb7ddcc330a5e3173372ea6af4a3d08ec58074478e85aa5603e926658",
                                    "compression_algorithm": "zlib"
                                },
                                "endpoint-exceptionlist-windows-v1": {
                                    "encryption_algorithm": "none",
                                    "decoded_sha256": "d801aa1fb7ddcc330a5e3173372ea6af4a3d08ec58074478e85aa5603e926658",
                                    "decoded_size": 14,
                                    "encoded_sha256": "f8e6afa1d5662f5b37f83337af774b5785b5b7f1daee08b7b00c2d6813874cda",
                                    "encoded_size": 22,
                                    "relative_url": "/api/endpoint/artifacts/download/endpoint-exceptionlist-windows-v1/d801aa1fb7ddcc330a5e3173372ea6af4a3d08ec58074478e85aa5603e926658",
                                    "compression_algorithm": "zlib"
                                }
                            }
                        }
                    },
                    "policy": {
                        "value": {
                            "windows": {
                                "events": {
                                    "dll_and_driver_load": true,
                                    "dns": true,
                                    "file": true,
                                    "network": true,
                                    "process": true,
                                    "registry": true,
                                    "security": true
                                },
                                "malware": {
                                    "mode": "prevent"
                                },
                                "logging": {
                                    "file": "info"
                                }
                            },
                            "mac": {
                                "events": {
                                    "process": true,
                                    "file": true,
                                    "network": true
                                },
                                "malware": {
                                    "mode": "prevent"
                                },
                                "logging": {
                                    "file": "info"
                                }
                            },
                            "linux": {
                                "events": {
                                    "process": true,
                                    "file": true,
                                    "network": true
                                },
                                "logging": {
                                    "file": "info"
                                }
                            }
                        }
                    }
                }
            }
        ],
        "package": {
            "name": "endpoint",
            "title": "Elastic Endpoint",
            "version": "0.8.0"
        },
        "revision": 1,
        "created_at": "2020-08-31T13:00:54.776Z",
        "created_by": "elastic",
        "updated_at": "2020-08-31T13:00:54.776Z",
        "updated_by": "elastic"
    },
    "success": true
}
```
I noticed that both windows and mac has the `policy.value.${OçS.malware.mode`, but Linux does not have. Is it correct? For the assertions, if we change from detect to prevent mode, should we check all platforms?

Reply:
> We could re-use the endpoint/metadata call POST, but I would like to exercise another API endpoint that is more direct:
GET to api/endpoint/metadata/[endpoint id]
>
>The endpoint id can be captured from the api/endpoint/metadata call we had just made (or we can make it again)

@EricDavisX about the endpoint id, where do we obtain it from? Maybe from metadata response, for the agentID, under this key `metadata.Endpoint.policy.applied.id`?

Reply:
No, I checked that an API call to that value returns a 404

Reply:
My understanding is that, after looping through the metadata response (hosts), we can get the status from there, right?